feat(AIP-111): introducing AIP around planes. (#1132)
googleapis/aip: aip/general/0111.md

---
id: 111
state: draft
created: 2023-06-10
placement:
  category: api-concepts
  order: 50
---

# Planes

Resources and methods on an API can be divided into the *plane* that they reside
or perform operations upon. For the context of APIs, the following planes are
defined:

> - The management plane: a uniform, resource-oriented API that primarily

Reviewer: We probably need to at least mention the control plane here, even if it may not be distinct from the management plane, since that term is likely more familiar. 

Also, some people may consider the IaC layer the management plane and the resource APIs to be the control plane.

Author: That begs the question: do we want to introduce a term like the management plane? It's a bit easier to define operations in terms of their impact on the control plane vs data-plane.

Although I guess this wouldn't capture that nuance that sometimes an operation may want to operate on both (e.g. control plane provisions a database and immediately populates it with a table or permissions).

> - The management plane: a uniform, resource-oriented API that primarily

Reviewer 2: Nit:
```suggestion
- Management plane: a uniform, resource-oriented API that primarily
```

  configures and allows retrieval of resources.

> - The data plane: a heteregenous API (ideally resource-oriented) that interface

Reviewer: This is missing the key aspect of the data plane, which is that it writes and reads user data. In the case of resource-oriented APIs, the data are associated with resources.

Reviewer: We may also want to include "and/or connects to provisioned entities, such as load balancers or virtual machines"

Author: rephrased to emphasize reading / writing of user data, connected to entities provisioned by the management plane. PTAL.

> - The data plane: a heteregenous API (ideally resource-oriented) that interface

Reviewer 2: Nit:
```suggestion
- Data plane: a heterogenous API (ideally resource-oriented) that interface
```

  with the resources that the the management plane configures.

The term "plane" was originally used in networking architecture. Although system
and network architecure often defines additional planes (e.g. control plane or
power planes), as the AIPs are focused on the interface, they are not defined in
this AIP.

## Guidance

### Management Plane

Management resources and methods exist primarily to provision, configure, and
audit the resources that the data plane interfaces with.

As an example, the following are considered management resources for a cloud
provider:

- virtual machines
- virtual private networks

> - disks

Reviewer: Virtual disks (for consistency)

- a blob store instance
- a project or account

The following behaviors are common (but not required) attributes of a
management plane resource:

> - changes infrequently.

Reviewer: This won't be true in the case of control loops, such as autoscaling, built on the standard resource APIs.

Reviewer 2: What does this even mean though? An instance of the resource doesn't change often? The definition of the resource doesn't change often? It's a tad vague imo

Reviewer: We need consistent terminology for referring to (a) the resource attributes, sometimes known as metadata, but that's ambiguous since fields like name, creation time, and annotations are also referred to as metadata, and (b) user data.

Author: for now I can remove this section - we can always add something to that effect later, and I think the "major distinctions" section does a good job highlighting key differences.

> - changes infrequently.

Reviewer 2: Nit:
```suggestion
- changes infrequently
```

### Data Plane

> Methods on the data plane operate on a variety of data formats, and generally

Reviewer: Methods on the data plane operate on user data in a variety of formats.

interface with a resource provisioned via a management plane API. Examples of
data plane methods include:

> - tables in a database

Reviewer: "tables" are not methods, which are verbs. They are nouns. Anyway, I wouldn't pick ambiguous examples. This is more about 3P vs 1P. Kubernetes is all control plane / management plane, but hosted 3P/OSS.

Author: Sure, I'd like to round back on these I can remove the example.

For 3P/OSS I'll follow up in the other thread.

Author: Also ack on tables not being methods. I guess we are looking at method categorization here.

> - rows in a table

Reviewer: writing and querying/reading rows in a table

Author: I'll leave it as writing or reading rows, since querying is implied. feel free to reopen if you think we need the querying part.

> - pushing or pulling from a message queue

Reviewer: pushing to

> - uploading or downloading blobs from a blob store instance

Reviewer 2: uploading blobs to

Data plane APIs **may** be heterogenous across a larger API surface, due to
requirements including high throughput, low latency, or the need to adhere to an
existing interface specification (e.g. ANSI SQL).

- For convenience, resources and methods that operate on the data plane **may**
expose themselves via resource-oriented management APIs. If so, those resources

> and methods **must** adhere to the requirements of the management plane.

Reviewer 3: What are `requirements of the management plane`?

Author: one is in AIP-133 around user-settable IDs management APIs should allow user-settable IDs on resources while data-plane does not.

I'd rather not enumerate them here because we have issues with maintaining duplicate guidance in multiple AIPs: https://github.com/aip-dev/google.aip.dev/pull/1139.

I added a sentence to note that the guidance is in other AIPs. WDYT?

Reviewer: I think that refers to the resource-oriented AIPs. It would be useful to refer to that explicitly.

Author: added!

### Major distinctions between management and data plane

- [IaC][] tools operate on the management plane exclusively.

> - Data planes provide much higher reliability, since they are often critical to

Reviewer 2: Data planes are much more performance sensitive than management planes.

Not sure if the broader "reliability" term was meant to address that.

Reviewer: I interpreted it as availability

Reviewer 2: Yeah, so did I. I think performance/high QPS capabilities are an important difference (we had a good discussion in our morning meeting about this, actually). Might be worth mentioning, but not blocking if it needs more discussion.

Author: Yes, availability might be a better term. I've added the points on both performance-sensitive and higher-throughput requirements.

  the runtime of applications.

### Planes are contextual

The definition of which plane a resource or operation resides on depends on the
context. Therefore, the definition of management or data plane is incomplete
without stating the context explicitly.

> For example, consider a cloud provider which exposes a Kubernetes cluster

Reviewer: I suppose this gets at the 1P vs hosted 3P issue. I would frame it more in those terms.

Author: It it always 1P vs 3P? for example, BigQuery is a 1P service, but certainly operations against the bigquery table are data-plane? 

I do agree that 1P vs 3P provides a clean delineation where applicable. That could be the first question to determine whether your operation is in the management or data plane.

Reviewer: Tables are resources: https://cloud.google.com/bigquery/docs/reference/rest/v2/tables

CRUDL operations on tables would be management plane.

If they were defined with SQL statements, that would be 3P.

Reading/writing data in the table is data plane.

Author: I think I'm going to remove this section - to your point it speaks more to 1st and third-party APIs, and we need a separate section on that one when defining the scope of the AIPs. 

There's already a section that talks about data-plane operations being exposed in a resource-oriented fashion.

resource, which in turn exposes it's own API with resources. In the context of
the cloud, the resources which live in the Kubernetes cluster are *data plane*.

> In the context of the Kubernetes cluster, the resources are *management plane*.

Reviewer 2: When you say "resources" in here do you mean methods? b.c AIP-121 resources aren't "data plane". In some of the "management plane resources" above, they are all nouns, not verbs. So I just want a little clarity here.

Reviewer: +1. This is about hosted 3P/OSS.

Author: sorry about that, I uploaded the change to make it about methods.

Author: removed the planes are contextual section entirely, will replace with a "1p vs 3p" discussion in the AIPs in a separate PR.

---
id: 111

> state: draft

Reviewer 2: When can we change it to approved?...before submitting? :D

Love this guidance
